src/Util/ComposeFileManipulator.php

```diff
@@ -0,0 +1,78 @@
+<?php
+
+/*
+ * This file is part of the Symfony MakerBundle package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\MakerBundle\Util;
+
+/**
+ * Manipulate Docker Compose Files.
+ *
+ * @author Jesse Rushlow <jr@rushlow.dev>
+ *
+ * @internal
+ * @final
+ */
+class ComposeFileManipulator
+{
+    public const COMPOSE_FILE_VERSION = '3.7';
+
+    private $composeData;
+
+    public function __construct(string $contents = '')
+    {
+        if ('' === $contents) {
+            $this->composeData = $this->getBasicStructure();
+        } else {
+            $this->composeData = (new YamlSourceManipulator($contents))->getData();
+        }
+    }
+
+    public function getComposeData(): array
+    {
+        return $this->composeData;
+    }
+
+    public function serviceExists(string $name): bool
+    {
+        if (\array_key_exists('services', $this->composeData)) {
+            return \array_key_exists($name, $this->composeData['services']);
+        }
+
+        return false;
+    }
+
+    public function addDockerService(string $name, array $details): void
+    {
+        $this->composeData['services'][$name] = $details;
+    }
+
+    public function removeDockerService(string $name): void
+    {
+        unset($this->composeData['services'][$name]);
+    }
+
+    public function exposePorts(string $service, array $ports): void
+    {
+        $this->composeData['services'][$service]['ports'] = $ports;
```

Comment:
Same here and below - we should probably include some `isset()` checks and throw an exception if something is wrong. It would be totally unexpected - but better than having an offset does not exist error.

Reply:
i agree - see above comments